eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+            trackingToken = newTrackingToken;
+
+            logger.debug("Message consumed from stream: {}", streamIdOfMessage);
+            return new GenericTrackedEventMessage<>(newTrackingToken, messageToReturn);
+        }
+
+        private void peekForMessages(HashMap<String, TrackedSourcedEventMessage> candidateMessagesToReturn) {
+            for (Map.Entry<String, BlockingStream<TrackedEventMessage<?>>> singleMessageSource : messageSources
+                    .entrySet()) {
+                Optional<TrackedEventMessage<?>> currentPeekedMessage = singleMessageSource.getValue().peek();
+                currentPeekedMessage.ifPresent(trackedEventMessage ->
+                                                       candidateMessagesToReturn.put(singleMessageSource.getKey(),
+                                                                                     new GenericTrackedSourcedEventMessage(
+                                                                                             singleMessageSource
+                                                                                                     .getKey(),
+                                                                                             trackedEventMessage)));
+            }
+        }
+
+        /**
+         * Calls close on each of the streams.
+         */
+        @Override
+        public void close() {
+            for (Map.Entry<String, BlockingStream<TrackedEventMessage<?>>> singleMessageSource : messageSources
+                    .entrySet()) {
+                singleMessageSource.getValue().close();
+            }
+        }
+    }
+
+
+    /**
+     * Builder class to instantiate a {@link MultiStreamableMessageSource}. The configurable filed
+     * {@code comparator}, which decides which message to process first if there is a choice defaults to the oldest
+     * message available (using the event's timestamp). The stream on which long polling is done for
+     * {@link MultiSourceMessageStream#hasNextAvailable(int, TimeUnit)} is also configurable.
+     */
+    public static class Builder {
+
+        private Comparator<TrackedSourcedEventMessage<?>> comparator = Comparator.comparing(EventMessage::getTimestamp);
+        private Map<String, StreamableMessageSource<TrackedEventMessage<?>>> messageSourceMap = new LinkedHashMap<>();
+        private String longPollingSource = "";
+
+        /**
+         * Adds a message source to the list of sources.
+         *
+         * @param messageSourceName a unique name identifying the stream.
+         * @param messageSource     the message source to be added.
+         * @return the current Builder instance, for fluent interfacing.
+         */
+        public Builder addMessageSource(String messageSourceName,
+                                        StreamableMessageSource<TrackedEventMessage<?>> messageSource) {
+            Assert.isFalse(messageSourceMap.containsKey(messageSourceName),
```

Comment:
I think it makes sense to utilize the `BuilderUtils` here instead of `Assert`, as those ensure we throw a `AxonConfigurationException`.